### PR TITLE
Implement authentication via VIISP

### PR DESCRIPTION
**Pull request by mjurkus**
also known as Elektroniniai Valdžios Vartai.

This authentication type will be used when submitting reports about parking violations. User will no longer be required to do that manually.

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -108,6 +108,8 @@
                 android:value=".MainActivity" />
         </activity>
 
+        <activity android:name=".activity.ViispLoginActivity" />
```
Comment (vycius) [inline app/src/main/AndroidManifest.xml]: Possibly, it's better to stick with portrait mode, set label and parent activity as in all other activities.

File: app/src/main/java/lt/vilnius/tvarkau/activity/ViispLoginActivity.kt [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    override fun onBackPressed() {
+        if (web_view.canGoBack()) {
+            web_view.goBack()
+        } else {
+            super.onBackPressed()
+        }
+    }
+
+    companion object {
+        private const val VIISP_AUTH_URI = "https://api.tvarkaumiesta.lt/auth/viisp/new"
```
Comment (vycius) [inline app/src/main/java/lt/vilnius/tvarkau/activity/ViispLoginActivity.kt]: I don't really hardcoding of API host URL. Host can be injected from `RestAdapterModule`.

File: app/src/main/java/lt/vilnius/tvarkau/auth/SessionTokenImpl.kt [hunk 3]
```diff
@@ -48,9 +48,24 @@ class SessionTokenImpl @Inject constructor(
...
+                .requestAccessToken {
+                    handleAccessToken(it, emitter)
+                }
+        }
+    }
+
     private fun handleAccessToken(response: OAuthResponse, emitter: CompletableEmitter) {
         if (response.isSuccessful) {
-            appPreferences.apiToken.set(gsonSerializer.fromJson(response.body, ApiToken::class.java))
+            appPreferences.apiToken.set(gsonSerializer.fromJson(response.body, ApiToken::class.java), commit = true)
```
Comment (vycius) [inline app/src/main/java/lt/vilnius/tvarkau/auth/SessionTokenImpl.kt]: Why ? Commit vs. apply should behave the same.

File: app/src/main/java/lt/vilnius/tvarkau/activity/LoginActivity.kt [hunk 4]
```diff
@@ -84,6 +89,7 @@ class LoginActivity : BaseActivity() {
     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
         when (requestCode) {
             RC_SIGN_IN -> viewModel.signInWithGoogle(data)
+            RC_VIISP_SIGN_IN -> viewModel.signInWithViisp(data)
```
Comment (vycius) [inline app/src/main/java/lt/vilnius/tvarkau/activity/LoginActivity.kt:92]: Don't we care about `resultCode` ? I think we should call this only on `RESULT_OK`.

**Review by vycius — APPROVED** [review]
Looks good. Left few minor comments.

File: app/src/main/java/lt/vilnius/tvarkau/auth/SessionTokenImpl.kt (hunk 3 above)
Comment (mjurkus) [reply app/src/main/java/lt/vilnius/tvarkau/auth/SessionTokenImpl.kt]: Nope. Apply is done async. And I had cases where user refresh was done milliseconds before token was set :)

File: app/src/main/java/lt/vilnius/tvarkau/activity/LoginActivity.kt (hunk 4 above)
Comment (mjurkus) [reply app/src/main/java/lt/vilnius/tvarkau/activity/LoginActivity.kt:92]: Well now we don't set any other results. Have any ideas when we should set other result codes?

File: app/src/main/java/lt/vilnius/tvarkau/auth/SessionTokenImpl.kt (hunk 3 above)
Comment (vycius) [reply app/src/main/java/lt/vilnius/tvarkau/auth/SessionTokenImpl.kt]: Interesting. I always thought that it should behave the same as SharedPreferences instances are singletons.

File: app/src/main/java/lt/vilnius/tvarkau/activity/LoginActivity.kt (hunk 4 above)
Comment (vycius) [reply app/src/main/java/lt/vilnius/tvarkau/activity/LoginActivity.kt:92]: Doesn't matter. It should be ok as we don't have any other `setResult` inside activity.

File: app/src/main/java/lt/vilnius/tvarkau/auth/SessionTokenImpl.kt (hunk 3 above)
Comment (vycius) [reply app/src/main/java/lt/vilnius/tvarkau/auth/SessionTokenImpl.kt]: Just looked at Android docs https://developer.android.com/reference/android/content/SharedPreferences.Editor#apply()
```
As SharedPreferences instances are singletons within a process, it's safe to replace any instance of commit() with apply() if you were already ignoring the return value.

You don't need to worry about Android component lifecycles and their interaction with apply() writing to disk. The framework makes sure in-flight disk writes from apply() complete before switching states.
```

**Review by vycius — APPROVED** [review]

**Pull request merged**